### DISCUSSION TITLE
How about moving sStatusStrings array into Flash? Save 12 bytes of RAM!

### DISCUSSION THREAD

**Discussion by brewmanz** [bxparks/AceRoutine#45] [Ideas]
It's possible to save 12 bytes of static RAM by moving `sStatusStrings` array into Flash. 
`HelloCoroutine.ino` (Nano, with extra `CoroutineScheduler::list(Serial);` in `printHelloWorld`) 
Initial Memory : 2820 Flash, 240 Global
Making changes to `Coroutine.h`
```
    void statusPrintTo(Print& printer) {
      printer.print((__FlashStringHelper*)pgm_read_word(&sStatusStrings[mStatus])); // add pgm_read_word etc
    }
```
Making changes to `Coroutine.cpp`
```
const __FlashStringHelper* const sStatusStrings[] PROGMEM = { // add PROGMEM
```
Flash memory: 2820 Flash, 228 Global
A saving of 12 whole bytes!

(PS Some other irrelevant local changes made for Serial baudrate etc, should Flash not exactly match official version)

**Comment by bxparks** [comment]
If I remember my thinking on this, the `sStatusString[]` is used only by `CoroutineScheduler::list()`. If the `list()` function is not called in a program, the linker will optimize away the unused `sStatusString[]` so the `PROGMEM` will make no difference. The `list()` function is meant only for debugging, not for production. But even for debugging, I never found it as useful as I thought it would be. With some of the recent changes to reduce the flash size of a `Coroutine`, the name of the coroutine is no longer stored (to save flash memory), so the `list()` function is even less useful. Coupled with the fact that on AVR processors with limited flash, I recommend not using the `CoroutineScheduler` to save flash memory, and instead calling `Coroutine::runCoroutine()` directly in the global `loop()` function.

So putting it all together, I didn't feel like the 10-14 bytes of savings was worth the hassle. Maybe if I feel motivated before the next release, I might look into this. I have to regenerate the entire suite of benchmarks (AutoBenchmark and MemoryBenchmark), which is not hard, but time consuming, even with all the automated scripting infrastructure that I have built myself. So I'm probably not going to do all the work, just for this.

**Reply by bxparks** [reply]
FYI, I moved the `sStatusString[]` array to `PROGMEM`, since I am about to push a release. Saves 12 bytes RAM (ignore the "14 bytes" in the commit message, I obviously can't count).

**Comment by brewmanz** [comment]
If it's just for Debug, and you want to save even more space ... use self-defining data!
One `enum` and `const` data techniques I use is to make it a number that is also a char (or C-string is another technique)
For example:
```
    static const Status kStatusSuspended = 'S'; // was 0;

    /** Coroutine returned using the COROUTINE_YIELD() statement. */
    static const Status kStatusYielding = 'Y'; // was 1;

    /** Coroutine returned using the COROUTINE_DELAY() statement. */
    static const Status kStatusDelaying = 'D'; // was 2;

    /** Coroutine is currenly running. True only within the coroutine itself. */
    static const Status kStatusRunning = 'R'; // was 3;

    /** Coroutine executed the COROUTINE_END() statement. */
    static const Status kStatusEnding = 'E'; // was 4;

    /** Coroutine has ended and no longer in the scheduler queue. */
    static const Status kStatusTerminated = 'T'; // was 5;
```
and write with
```
      printer.print((char)mStatus);
```
And the C-String (only up to 3 printable chars, and better suited where RAM is not an issue)
```
    typedef uint32_t Status;

    static const Status kStatusSuspended = 'S' + 'u'*0x100 + 's'*0x10000; // was 0;

    /** Coroutine returned using the COROUTINE_YIELD() statement. */
    static const Status kStatusYielding = 'Y' + 'l'*0x100 + 'd'*0x10000; // was 1;

    /** Coroutine returned using the COROUTINE_DELAY() statement. */
    static const Status kStatusDelaying = 'D' + 'l'*0x100 + 'y'*0x10000; // was 2;

    /** Coroutine is currenly running. True only within the coroutine itself. */
    static const Status kStatusRunning = 'R' + 'u'*0x100 + 'n'*0x10000; // was 3;

    /** Coroutine executed the COROUTINE_END() statement. */
    static const Status kStatusEnding = 'E' + 'n'*0x100 + 'd'*0x10000; // was 4;

    /** Coroutine has ended and no longer in the scheduler queue. */
    static const Status kStatusTerminated = 'T' + 'r'*0x100 + 'm*0x10000'; // was 5;
```
and write with
```
      printer.print((const char*)&mStatus);
```
(You can probably guess that I am an Assembler programmer at heart)

**Reply by bxparks** [reply]
Ha, interesting idea on the encoding into `uint32_t`. I've done things like that for magic number in file formats. But for this? Umm, no. It actually wouldn't be so bad with the help of some CPP macros, but it's not worth the extra complexity just for this.
